gossipsub.go

```diff
@@ -238,17 +351,44 @@ func (gs *GossipSubRouter) handleIWant(p peer.ID, ctl *pb.ControlMessage) []*pb.
 
 func (gs *GossipSubRouter) handleGraft(p peer.ID, ctl *pb.ControlMessage) []*pb.ControlPrune {
 	var prune []string
+
+	doPX := true
+	score := gs.score.Score(p)
+
 	for _, graft := range ctl.GetGraft() {
 		topic := graft.GetTopicID()
 		peers, ok := gs.mesh[topic]
 		if !ok {
+			// don't do PX when there is an unknown topic to avoid leaking our peers
+			doPX = false
+			// spam harndening: ignore GRAFTs for unknown topics
+			continue
+		}
+
+		// check the score
+		if score < 0 {
```

Comment:
In general, from a security perspective, I would prefer if all the `score` validation code would happen at a single point, similar to what is being done in `AcceptFrom`. This is not a blocker for me because I understand that it's not as straightforward as that other case, here different handlers will react slightly differently to a blacklisted (or similar) peer.

Citing this particular line as an example of the rationale behind this comment (but this logic replicates in other places). If `score < 0` then we basically won't be doing any GRAFT or PX (seems we won't be doing much besides the prune control response). This is not at all evident buried here, and it makes it harder to audit and maintain, even if it's correct (which seems to be).

Ideally handlers like these one should deal with the positive logic, what actually gets done for accepted peers, and move all security checks upstream. Again, just a suggestion because there may be other repercussions for the proposed change I'm not seeing.

Reply:
it's a matter of some code duplication and convenience, but we can certainly check upfront here.

Reply:
ok, I tried it and the code gets convoluted and requires a goto. Not a very good tradeoff.

Reply:
@vyzo Why would you need a goto? Could you push that code in another branch please? I think I may have not expressed this correctly and we may thinking of different refactorings.